Code-related anti-patterns added (Documentation) (#12776)
elixir-lang/elixir: lib/elixir/pages/anti-patterns/code-anti-patterns.md

## Complex branching

#### Problem

> When a function assumes the responsibility of handling multiple errors alone, it can increase its cyclomatic complexity (metric of control-flow) and become incomprehensible. This situation can configure a specific instance of "Long function", a traditional anti-pattern, but has implications of its own. Under these circumstances, this function could get very confusing, difficult to maintain and test, and therefore bug-proneness.

Reviewer: ```suggestion
When a function assumes the responsibility of handling multiple errors alone, it can increase its cyclomatic complexity (metric of control-flow) and become incomprehensible. This situation can configure a specific instance of "Long function", a traditional anti-pattern, but has implications of its own. Under these circumstances, this function could get very confusing, difficult to maintain and test, and therefore bug-prone.
```

Reviewer: Oh, sad to see this one didn't get in. The context is "this function... bug-prone." Bug-proneness would be in the case of writing "bug-proneness would follow", for example.

#### Example

An example of this anti-pattern is when a function uses the `case` control-flow structure or other similar constructs (for example, `cond`, or `receive`) to handle multiple variations of response types returned by the same API endpoint. This practice can make the function more complex, long, and difficult to understand, as shown next.

```elixir
def get_customer(customer_id) do
  case get("/customers/#{customer_id}") do
    {:ok, %Tesla.Env{status: 200, body: body}} -> {:ok, body}
    {:ok, %Tesla.Env{body: body}} -> {:error, body}
    {:error, _} = other -> other
  end
end
```

> Although `get_customer/1` is not really long in this example, it could be. Thinking about this more complex scenario, where a large number of different responses can be provided to the same endpoint, is not a good idea to concentrate all on a single function. This is a risky scenario, where a little typo, or any problem introduced by the programmer in handling a response type, could eventually compromise the handling of all responses from the endpoint (if the function raises an exception, for example).

Reviewer: ```suggestion
Although `get_customer/1` is not really long in this example, it could be. Thinking about this more complex scenario, where a large number of different responses can be provided to the same endpoint, it's not a good idea to concentrate all on a single function. This is a risky scenario, where a little typo, or any problem introduced by the programmer in handling a response type, could eventually compromise the handling of all responses from the endpoint (if the function raises an exception, for example).
```

> Although `get_customer/1` is not really long in this example, it could be. Thinking about this more complex scenario, where a large number of different responses can be provided to the same endpoint, is not a good idea to concentrate all on a single function. This is a risky scenario, where a little typo, or any problem introduced by the programmer in handling a response type, could eventually compromise the handling of all responses from the endpoint (if the function raises an exception, for example).

Reviewer 2: I find this example doesn't illustrate the point really well, because in this precise case the "problematic" code looks clearer and more idiomatic to me 😅

How about making it look obviously problematic? We could use some ellipsis like this:

```elixir
    {:ok, %Tesla.Env{status: 200, body: body}} ->
      # complex processing done directly in the clause
      processed_response = ...
      {:ok, processed_response}
```

Reviewer 3: Agreed. We need to say that each clause is made by several lines of code. The current example does not portray that. :)

Author: I have a feeling that employing ellipses in this instance would further abstract the example. What exactly constitutes "complex processing"? I would appreciate a more tangible code example that effectively illustrates the concept, but without making the documentation long and tiresome to read. :)

Reviewer 2: How about something like

```elixir
body
|> Jason.decode!()
|> Map.fetch!("key")
|> String.trim()
```

Reviewer 3: @fuelen we do cover more general anti-patterns, such as long parameters, so this one would fit. You are right, it applies to other cases, such as `if/else`, so perhaps that's worth a mention. It may also be worth mentioning that Elixir scopes make it _very_ easy to refactor. I can add some pseudo json-decoding to the function and refactor it from there.

Reviewer 4: In my previous response, I wanted to mention that the function size is not so crucial for immutable languages, and moving code to a separate function doesn't always result in better code.

I suggest thinking about a different example. The current one exposes HTTP request details to the end user - in case of an error, Tesla can return `{:error, %Tesla.Error{}}` and the error contains HTTP request metadata. I think this is an antipattern as well. In the real world, people will use a middleware provided by Tesla to decode JSON.
Also, I think it is better to avoid the usage of 3rd-party libraries in these docs, as people will treat this as an officially-recommended library.

PS: I see there was a discussion about the previous version of the smell https://github.com/lucasvegi/Elixir-Code-Smells/issues/4 The original one had a smell mentioned here https://keathley.io/blog/good-and-bad-elixir.html in `Don’t pipe results into the following function` section

Reviewer 3: Yup, I will refactor the Tesla part as well. 

>  I wanted to mention that the function size is not so crucial for immutable languages, 

That's why we removed the function size anti-pattern because it is hard to provide a specific rule for this. But, for complex branching, there is an easier argument to be made that the code is harder to grasp when the branches themselves are so far apart that you cannot see them at a glance.

So I will be going ahead with this one, including the changes above. :)

#### Refactoring

> As shown below, in this situation, instead of concentrating all handlings within the same function, creating a complex branching, it is better to delegate each branch (handling of a response type) to a different private function. In this way, the code will be cleaner, more concise, and readable.

Reviewer: ```suggestion
As shown below, in this situation, instead of concentrating all handlings within the same function, thus creating more complex branching, it is better to delegate each branch (handling of a response type) to a different private function. In this way, the code will be cleaner, more concise, and readable.
```

Reviewer: I'm also not sure if `handlings` wouldn't be better expressed as `handlers`.

Reviewer: Hm... this was resolved without a discussion or follow-up, but the proposed changes didn't make it. "a complex branching" seems incorrect, since the "a" pertains to something, which would eventually follow branching...

```elixir
def get_customer(customer_id) when is_integer(customer_id) do
  case get("/customers/#{customer_id}") do
    {:ok, %Tesla.Env{status: 200, body: body}} -> success_api_response(body)
    {:ok, %Tesla.Env{body: body}} -> x_error_api_response(body)
    {:error, _} = other -> y_error_api_response(other)
  end
end

defp success_api_response(body) do
  {:ok, body}
end

defp x_error_api_response(body) do
  {:error, body}
end

defp y_error_api_response(other) do
  other
end
```

> While this example of refactoring `get_customer/1` might seem quite more verbose than the original code, remember to imagine a scenario where `get_customer/1` is responsible for handling a number much larger than three different types of possible responses. This is the smelly scenario!

Reviewer: ```suggestion
While this example of refactoring `get_customer/1` might seem quite more verbose than the original code, imagine a scenario where `get_customer/1` is responsible for handling a number much larger than three different types of possible responses. This is the smelly scenario!
```

## Complex `else` clauses in `with`

#### Problem

> This anti-pattern refers to `with` statements that flatten all its error clauses into a single complex `else` block. This situation is harmful to the code readability and maintainability because difficult to know from which clause the error value came.

Reviewer: ```suggestion
This anti-pattern refers to `with` statements that flatten all its error clauses into a single complex `else` block. This situation is harmful to the code readability and maintainability because it makes it difficult to know from which clause the error value came.
```

#### Example

> An example of this anti-pattern, as shown below, is a function `open_decoded_file/1` that read a base 64 encoded string content from a file and returns a decoded binary string. This function uses a `with` statement that needs to handle two possible errors, all of which are concentrated in a single complex `else` block.

Reviewer: ```suggestion
An example of this anti-pattern, as shown below, is function `open_decoded_file/1`, that read a Base64-encoded string from a file and returns a decoded binary string. This function uses a `with` statement that needs to handle two possible errors, all of which are concentrated in a single complex `else` block.
```

```elixir
def open_decoded_file(path) do
  with {:ok, encoded} <- File.read(path),
    {:ok, value} <- Base.decode64(encoded) do
    value
  else
    {:error, _} -> :badfile
    :error -> :badencoding
  end
end
```

#### Refactoring

> In this situation, instead of concentrating all error handlings within a single complex `else` block, it is better to normalize the return types in specific private functions. In this way, due to its organization, the code will be cleaner and more readable.

Reviewer: Same comment for `handlings` as before. It's possible it's intended, but it's a term I've not seen before in similar contexts.

Reviewer 3: "all error clauses" or "all error handling". I prefer the former but I am fine either way.

```elixir
def open_decoded_file(path) do
  with {:ok, encoded} <- file_read(path),
    {:ok, value} <- base_decode64(encoded) do
    value
  end
end

defp file_read(path) do
  case File.read(path) do
    {:ok, contents} -> {:ok, contents}
    {:error, _} -> :badfile
  end
end

defp base_decode64(contents) do
  case Base.decode64(contents) do
    {:ok, contents} -> {:ok, contents}
    :error -> :badencoding
  end

> end

Reviewer 4: I think such output types as `{:ok, binary()} | :badfile` and `{:ok, binary()} | :badencoding` are not conventional and it is better to rewrite them to `{:ok, binary()} | {:error, :badfile}` and `{:ok, binary()} | {:error, :badencoding}`
